src/fklearn/validation/evaluators.py

```diff
@@ -603,6 +603,73 @@ def spearman_evaluator(test_data: pd.DataFrame,
     return {eval_name: score}
 
 
+@curry
+def ndcg_evaluator(test_data: pd.DataFrame,
+                   prediction_column: str = "prediction",
+                   target_column: str = "target",
+                   k: int = None,
+                   exponential_gain: bool = True,
+                   eval_name: str = None) -> EvalReturnType:
+    """
+    Computes the Normalized Discount Cumulative Gain (NDCG) between
+    of the original and predicted rankings:
+    https://en.wikipedia.org/wiki/Discounted_cumulative_gain
+
+    Parameters
+    ----------
+
+    test_data : Pandas DataFrame
+        A Pandas' DataFrame with with target and prediction scores.
+
+    prediction_column : String
+        The name of the column in `test_data` with the prediction scores.
+
+    target_column : String
+        The name of the column in `test_data` with the target.
+
+    k : int, optional (default=None)
+        The size of the rank that is used to fit (highest k scores) the NDCG score. If None, use all outputs.
+        Note that this value must between [1, +inf].
+
+    exponential_gain : bool (default=True)
+        If False, then use the linear gain. The exponential gain places a stronger emphasis on retrieving
+        relevant items. If the relevance of these items is binary values in {0,1}, then the two approaches
+        are the same, which is the linear case.
+
+    eval_name : String, optional (default=None)
+        The name of the evaluator as it will appear in the logs.
+
+    Returns
+    ----------
+    log: dict
+        A log-like dictionary with the NDCG score.
+    """
+
+    if eval_name is None:
+        eval_name = f"ndcg_evaluator__{target_column}"
+
+    if isinstance(k, int) and k <= 0:
```

Comment:
also, there's no representation for inf as integer in python, and the metric is not defined for `k` bigger than the size of the array. why not `0 <= k <= len(predictions)`?

Reply:
```
if not isinstance(k, int) or k < 1 or k > len(test_data[prediction_column])
```

Reply:
If we pass `k = 0`, then we get `RuntimeWarning: invalid value encountered in double_scalars`, because the rankings will be empty and as a result a `nan` score. Look that code snippet below:
```Python
rel = np.argsort(test_data[prediction_column])[::-1][:k]
```
To avoid this, I put this check. But now I think this is unnecessary and the `RuntimeWarning` is sufficient.

If `k > len(test_data[prediction_column])`, nothing happens! It's the same case when `k=None`. So, we can write something like as:

```Python
if isinstance(k, int) and k < 0:
        raise AttributeError("k must be a positive integer.")
```
The objective of this check is to verify if `k` is a positive integer number. If `None` (default value) then use all outputs to fit the NDCG score. The case when `k < 0` produces wrong scores.

Do you agree? Thanks!

Reply:
prefer explicit over implicit: why should we wait until code execution reaches this point and raises a runtime error? we can catch it much sooner.

Reply:
i would rely on implicit runtime assumptions: it's extremely cheap to write condition explicitly, i see no reason not do it

Reply:
I considered this because of your comment _"is this check correct? passing k=0.0 won't raise an error"_. But I can change to:

```Python3
if isinstance(k, int) and k <= 0:
    raise AttributeError("k must be a positive integer.")
```

Reply:
Think about the user: he calls this function and unexpectedly reads this in their stacktrace: RuntimeWarning: invalid value encountered in double_scalars`. What is more readable, this or simply adding one more check in this condition, and the user reading `k must be a positive integer`? Please don't rely on runtime errors — think about improving the user experience. It's pratically free changing this to: `not isinstance(k, int) or k < 1`.

Reply:
I totally agree with you, but this `not isinstance(k, int) or k < 1` is wrong! In the left side of the verification, I only need to verify if this is a number. If `None` then use all values of the rankings. This warning `RuntimeWarning: invalid value encountered in double_scalars` is **only** generated if `k = 0`. I understand the issues that you can solve with your check. But look at the parameters description. If `k = None`, is the default value to use all ratings to fit the NDCG score. I can change to:

```Python3
if isinstance(k, int) and k < 1:
    raise AttributeError("k must be a positive integer.")
```

Reply:
Sorry for the flow interruption, but the @gppeixoto 's point is super valid!
Your `if` statement should be:
```python
if isinstance(k, int, double) and k < 1:
    pass
```
Or better:
```python
assert isinstance(k, int, double) and k < 1, "k must be a positive integer."
assert k > len(test_data[prediction_column]), "k must be less or equal than dataset length"
```

Another point is, change the `AttributeError` to something more meaningful. If you read the description of this you'll see the real proposal of it, https://docs.python.org/3/library/exceptions.html#AttributeError.

Reply:
Sorry guys, I thought that the @gppeixoto's was related to only zero (not float type) assignment, sorry! Passing a float value won't raise an error. I'll make the changes as soon as possible. I'll follow @phakiller's suggestions.

```Python3
assert isinstance(k, (int, float)) and k < 1, "k must be a positive."
```
Thanks.

Reply:
no worries — thanks to @phakiller for the thorough suggestions!

feel free to @ me whenever you want me to take a look at it again